controllers/s3utils.go

```diff
@@ -252,6 +255,93 @@ func (s *s3ObjectStore) createBucket(bucket string) (err error) {
 	return nil
 }
 
+// deleteBucket deletes the S3 bucket.  Fails to delete if the bucket contains
+// any objects.
+func (s *s3ObjectStore) deleteBucket(bucket string) (
+	err error) {
+	if bucket == "" {
+		return fmt.Errorf("empty bucket name for "+
+			"endpoint %s caller %s", s.s3Endpoint, s.callerTag)
+	}
+
+	defer func() {
+		if r := recover(); r != nil {
+			// change the named return err value
+			err = fmt.Errorf("delete bucket recovered for %s, with %v",
+				bucket, r)
+		}
+	}()
+
+	dbInput := &s3.DeleteBucketInput{Bucket: &bucket}
+	if err = dbInput.Validate(); err != nil {
+		return fmt.Errorf("delete bucket input validation failed for %s, err %w",
+			bucket, err)
+	}
+
+	_, err = s.client.DeleteBucket(dbInput)
+	if err != nil {
+		var aerr awserr.Error
+		if errorswrapper.As(err, &aerr) {
+			switch aerr.Code() {
+			case s3.ErrCodeNoSuchBucket:
+			default:
+				return fmt.Errorf("failed to create bucket %s, %w",
+					bucket, err)
+			}
+		}
+	}
+
+	return nil
+}
+
+// purgeBucket empties the content of the given bucket.
+func (s *s3ObjectStore) purgeBucket(bucket string) (
+	err error) {
+	if bucket == "" {
+		return fmt.Errorf("empty bucket name for "+
+			"endpoint %s caller %s", s.s3Endpoint, s.callerTag)
+	}
+
+	defer func() {
```

Comment:
Good question. Given that the S3 inputs like endpoint, region, etc., come from the DR Policy spec, if something is wrong with it, it could crash the s3 client library. While I don't recall the exact scenario, I recall that I added this defer-panic-recover protective pattern when I ran into a panic with the create bucket call. I reused it for delete and purge bucket methods.